arch/xtensa/src/esp32/esp32_spi_slave.c

```diff
@@ -548,36 +754,78 @@ static int esp32_io_interrupt(int irq, void *context, FAR void *arg)
 
 static int esp32_spislv_interrupt(int irq, void *context, FAR void *arg)
 {
-  uint32_t rd;
-  uint32_t wd;
   struct esp32_spislv_priv_s *priv = (struct esp32_spislv_priv_s *)arg;
-  irqstate_t flags;
-
-  flags = enter_critical_section();
+  uint32_t n;
+  uint32_t tmp;
+  int i;
 
   esp32_spi_reset_regbits(priv, SPI_SLAVE_OFFSET, SPI_TRANS_DONE_M);
 
-  if (!priv->process)
+  if (priv->dma_chan)
+    {
+      esp32_spi_set_reg(priv, SPI_DMA_CONF_OFFSET, SPI_DMA_RESET_MASK);
+      esp32_spi_reset_regbits(priv, SPI_DMA_CONF_OFFSET,
+                              SPI_DMA_RESET_MASK);
+    }
+
+  if (priv->process == false)
     {
       SPI_SDEV_SELECT(priv->sdev, true);
       priv->process = true;
     }
 
-  rd = esp32_spi_get_reg(priv, SPI_W0_OFFSET);
-  if (spi_dequeue(priv, &wd))
+  /* Read and calculate read bytes */
+
+  n = (esp32_spi_get_reg(priv, SPI_SLV_RD_BIT_OFFSET) + 1) / 8;
+
+  esp32_spi_set_regbits(priv, SPI_USER_OFFSET, SPI_USR_MOSI_M);
+
+  /* RX process */
+
+  if (!priv->dma_chan)
     {
-      esp32_spi_set_reg(priv, SPI_W0_OFFSET, wd);
-      esp32_spi_set_regbits(priv, SPI_CMD_OFFSET, SPI_USR_M);
+      /** With DMA, software should copy data from regitser
+       *  to receive buffer
+       */
+
+      for (i = 0; i < n; i += 4)
+        {
+          tmp = esp32_spi_get_reg(priv, SPI_W0_OFFSET + i);
+          memcpy(priv->rxbuffer + priv->rxlen + i, &tmp, n);
+        }
     }
 
-  SPI_SDEV_RECEIVE(priv->sdev, &rd, 1);
+  priv->rxlen += n;
 
-  if (priv->process == false)
+  esp32_spislv_rx(priv);
+
+  /* TX process */
+
+  if (priv->txen)
     {
-      SPI_SDEV_SELECT(priv->sdev, false);
+      if (n < priv->txlen)
+        {
+          priv->txlen -= n;
```

Comment:
to me, it looks weird to use the same `n` value read from SPI_SLV_RD_BIT_OFFSET for both of rx and tx.
is this intended?